// orama/lib/Layer/lines/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lines = exports.hoverSolver = exports.getPointData = undefined;

var _isArray2 = require('lodash/isArray');

var _isArray3 = _interopRequireDefault(_isArray2);

var _some2 = require('lodash/some');

var _some3 = _interopRequireDefault(_some2);

var _reduce2 = require('lodash/reduce');

var _reduce3 = _interopRequireDefault(_reduce2);

var _head2 = require('lodash/head');

var _head3 = _interopRequireDefault(_head2);

var _isEmpty2 = require('lodash/isEmpty');

var _isEmpty3 = _interopRequireDefault(_isEmpty2);

var _last2 = require('lodash/last');

var _last3 = _interopRequireDefault(_last2);

var _findIndex2 = require('lodash/findIndex');

var _findIndex3 = _interopRequireDefault(_findIndex2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _find2 = require('lodash/find');

var _find3 = _interopRequireDefault(_find2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // Copyright 2017 Kensho Technologies, Inc.

var _path2DUtils = require('../../utils/path2DUtils');

var _utils = require('../../utils');

var _plotValue = require('../../Layer/plotValue');

var _getPlotValues = require('../../Layer/getPlotValues');

var _splineInterpolation = require('../../Layer/splineInterpolation');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
`points` is used to generate render data for lines and multilines.
it handles `x`, `y`, 'stroke'(color) and 'lineWisdth'.

@calling logic
lines{
  getLineRenderData{}
}
*/

var getPointData = exports.getPointData = function getPointData(props, datum) {
  var path2D = (0, _path2DUtils.getPath2D)();
  var x = (0, _plotValue.plotValue)(props, datum, undefined, 'x');
  var y = (0, _plotValue.plotValue)(props, datum, undefined, 'y');
  var r = (0, _plotValue.plotValue)(props, datum, undefined, 'strokeWidth', 2) + 1.5;
  if ((0, _utils.notPlotNumber)([x, y, r])) return undefined;
  path2D.arc(x, y, r, 0, 2 * Math.PI);
  return {
    hoverAlpha: 0.8,
    path2D: path2D,
    type: 'area'
  };
};

var getHoverSolverObj = function getHoverSolverObj(props, renderDatum, hoverData) {
  return {
    hoverRenderData: [renderDatum, getPointData(props, hoverData)],
    renderDatum: renderDatum,
    hoverData: hoverData
  };
};

var hoverSolver = exports.hoverSolver = function hoverSolver(props, _hoverData, renderDatum, localMouse) {
  var xRaw = props.xScale.invert(localMouse.x);
  if (props.xType === 'ordinal') {
    var _hoverData2 = (0, _find3.default)(_hoverData, function (d) {
      return (0, _get3.default)(d, props.x) === xRaw;
    });
    return getHoverSolverObj(props, renderDatum, _hoverData2);
  }
  var hoverIndex = (0, _findIndex3.default)(_hoverData, function (d) {
    return (0, _get3.default)(d, props.x) > xRaw;
  });
  if (hoverIndex === 0) {
    var _hoverData3 = _hoverData[hoverIndex];
    return getHoverSolverObj(props, renderDatum, _hoverData3);
  }
  if (hoverIndex === -1) {
    var _hoverData4 = (0, _last3.default)(_hoverData);
    return getHoverSolverObj(props, renderDatum, _hoverData4);
  }
  var px = (0, _get3.default)(_hoverData[hoverIndex], props.x);
  var x = (0, _get3.default)(_hoverData[hoverIndex - 1], props.x);
  if (xRaw - px < x - xRaw) {
    var _hoverData5 = _hoverData[hoverIndex - 1];
    return getHoverSolverObj(props, renderDatum, _hoverData5);
  }
  var hoverData = _hoverData[hoverIndex];
  return getHoverSolverObj(props, renderDatum, hoverData);
};

/*
generates the array of render data
*/
var getLineRenderData = function getLineRenderData(props, data, idx) {
  if ((0, _isEmpty3.default)(data)) return undefined;
  var path2D = (0, _path2DUtils.getPath2D)();
  var values = (0, _getPlotValues.getPlotValues)(props, (0, _head3.default)(data), idx, {
    hoverAlpha: 0.2
  });
  if (props.interpolate) {
    (0, _splineInterpolation.splineInterpolation)(props, data, path2D);
  } else {
    path2D.moveTo(values.x, values.y);
    (0, _reduce3.default)(data, function (shouldDrawPoint, d) {
      var x = (0, _plotValue.plotValue)(props, d, idx, 'x');
      var y = (0, _plotValue.plotValue)(props, d, idx, 'y');
      if ((0, _utils.notPlotNumber)([x, y])) return false;
      if (shouldDrawPoint) path2D.lineTo(x, y);else path2D.moveTo(x, y);
      return true;
    }, true);
  }
  return _extends({}, values, {
    data: data,
    hoverSolver: hoverSolver,
    path2D: path2D,
    type: 'line'
  });
};
var lines = exports.lines = function lines(props) {
  if (!props.xScale || !props.yScale) return undefined;
  if ((0, _some3.default)(props.data, _isArray3.default)) {
    return (0, _reduce3.default)(props.data, function (acc, data, idx) {
      return acc.concat(getLineRenderData(props, data, idx));
    }, []);
  }
  return [getLineRenderData(props, props.data)];
};
