// orama/lib/Chart/ChartRender/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartRender = undefined;

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _basicRender = require('../../CanvasRender/basicRender');

var _CanvasRender = require('../../CanvasRender');

var _defaultTheme = require('../../defaultTheme');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
Used inside <Chart/>
*/
var ChartRender = exports.ChartRender = function ChartRender(props) {
  return _react2.default.createElement(
    'div',
    null,
    (0, _map3.default)(props.renderLayers, function (renderLayer, i) {
      return _react2.default.createElement(_CanvasRender.CanvasRender // basicRender
      , { clip: true,
        height: props.rootProps.height,
        key: i,
        layerProps: renderLayer.layerProps,
        plotRect: props.rootProps.plotRect,
        render: _basicRender.basicRender,
        renderData: renderLayer.renderData,
        theme: props.theme,
        width: props.rootProps.width
      });
    })
  );
}; // Copyright 2017 Kensho Technologies, Inc.

ChartRender.propTypes = {
  renderLayers: _propTypes2.default.array,
  rootProps: _propTypes2.default.object,
  theme: _propTypes2.default.object
};
ChartRender.defaultProps = {
  theme: _defaultTheme.DEFAULT_THEME
};
