// orama/lib/Tooltip/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tooltip = exports.TooltipInner = undefined;

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // Copyright 2017 Kensho Technologies, Inc.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _defaultTheme = require('../defaultTheme');

var _extractTooltipData = require('../Chart/extractTooltipData');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MAX_WIDTH = 320;

var getPadding = function getPadding(props) {
  return props.theme.tooltipFontSize / 2;
};

var row = function row(props, d, i) {
  return _react2.default.createElement(
    'tr',
    {
      style: {
        background: i % 2 ? props.theme.tooltipBackgroundFill : props.theme.tooltipEvenBackgroundFill
      },
      key: i
    },
    props.showKeys ? _react2.default.createElement(
      'td',
      {
        style: {
          borderRight: '2px solid ' + props.theme.tooltipKeyBorderStroke,
          padding: getPadding(props)
        }
      },
      d.key
    ) : null,
    _react2.default.createElement(
      'td',
      { // Name
        style: {
          padding: getPadding(props),
          textAlign: 'left',
          verticalAlign: 'top'
        }
      },
      d.name
    ),
    _react2.default.createElement(
      'td',
      { // Value
        style: {
          fontFamily: props.theme.fontFamilyMono,
          fontSize: props.theme.tooltipValueFontSize,
          padding: getPadding(props),
          textAlign: 'right',
          verticalAlign: 'top'
        }
      },
      d.value
    )
  );
};
row.propTypes = {
  showKeys: _propTypes2.default.bool,
  theme: _propTypes2.default.object
};

var TooltipInner = exports.TooltipInner = function TooltipInner(props) {
  return _react2.default.createElement(
    'div',
    {
      style: {
        background: props.theme.tooltipBackgroundFill,
        boxShadow: '1px 1px 1px ' + props.theme.tooltipBoxShadowFill,
        color: props.theme.tooltipTextFill,
        fontFamily: props.theme.fontFamily,
        fontSize: props.theme.tooltipFontSize,
        maxWidth: MAX_WIDTH,
        opacity: 0.96
      }
    },
    props.title ? _react2.default.createElement(
      'div',
      {
        style: {
          fontSize: props.theme.tooltipTitleFontSize,
          fontWeight: props.theme.tooltipTitleFontWeight,
          padding: getPadding(props),
          textAlign: 'left',
          verticalAlign: 'top'
        }
      },
      props.title
    ) : null,
    _react2.default.createElement(
      'table',
      {
        style: { width: '100%', borderCollapse: "collapse" }
      },
      _react2.default.createElement(
        'tbody',
        null,
        (0, _map3.default)(props.values, function (d, i) {
          return row(props, d, i);
        })
      )
    )
  );
};
TooltipInner.propTypes = {
  mouse: _propTypes2.default.object,
  showKeys: _propTypes2.default.bool,
  theme: _propTypes2.default.object,
  title: _propTypes2.default.string,
  values: _propTypes2.default.array
};
TooltipInner.defaultProps = {
  theme: _defaultTheme.DEFAULT_THEME,
  showKeys: false
};

var Tooltip = exports.Tooltip = function Tooltip(props) {
  var tooltipData = (0, _extractTooltipData.extractTooltipData)(props.layerProps, props.hoverData);
  return _react2.default.createElement(TooltipInner, _extends({
    showKeys: props.layerProps.tooltipShowKeys,
    theme: props.theme
  }, tooltipData));
};
Tooltip.propTypes = {
  hoverData: _propTypes2.default.oneOfType([_propTypes2.default.array, _propTypes2.default.object]),
  layerProps: _propTypes2.default.object,
  theme: _propTypes2.default.object
};
