// orama/lib/Layer/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScatterplotLabels = exports.Text = exports.Ranges = exports.Points = exports.Lines = exports.Guides = exports.Brushes = exports.Bars = exports.Areas = exports.Layer = exports.getPlotValues = exports.plotValue = exports.text = exports.ranges = exports.points = exports.lines = exports.guides = exports.brushes = exports.bars = exports.areas = undefined;

var _areas = require('../Layer/areas');

Object.defineProperty(exports, 'areas', {
  enumerable: true,
  get: function get() {
    return _areas.areas;
  }
});

var _bars = require('../Layer/bars');

Object.defineProperty(exports, 'bars', {
  enumerable: true,
  get: function get() {
    return _bars.bars;
  }
});

var _brushes = require('../Layer/brushes');

Object.defineProperty(exports, 'brushes', {
  enumerable: true,
  get: function get() {
    return _brushes.brushes;
  }
});

var _guides = require('../Layer/guides');

Object.defineProperty(exports, 'guides', {
  enumerable: true,
  get: function get() {
    return _guides.guides;
  }
});

var _lines = require('../Layer/lines');

Object.defineProperty(exports, 'lines', {
  enumerable: true,
  get: function get() {
    return _lines.lines;
  }
});

var _points = require('../Layer/points');

Object.defineProperty(exports, 'points', {
  enumerable: true,
  get: function get() {
    return _points.points;
  }
});

var _ranges = require('../Layer/ranges');

Object.defineProperty(exports, 'ranges', {
  enumerable: true,
  get: function get() {
    return _ranges.ranges;
  }
});

var _text = require('../Layer/text');

Object.defineProperty(exports, 'text', {
  enumerable: true,
  get: function get() {
    return _text.text;
  }
});

var _plotValue = require('../Layer/plotValue');

Object.defineProperty(exports, 'plotValue', {
  enumerable: true,
  get: function get() {
    return _plotValue.plotValue;
  }
});

var _getPlotValues = require('../Layer/getPlotValues');

Object.defineProperty(exports, 'getPlotValues', {
  enumerable: true,
  get: function get() {
    return _getPlotValues.getPlotValues;
  }
});

var _scatterplotLabels = require('../Layer/scatterplotLabels');

var Layer = exports.Layer = function Layer() {
  return null;
};
var Areas = exports.Areas = function Areas() {
  return null;
};
Areas.plot = _areas.areas;
var Bars = exports.Bars = function Bars() {
  return null;
};
Bars.plot = _bars.bars;
var Brushes = exports.Brushes = function Brushes() {
  return null;
};
Brushes.plot = _brushes.brushes;
var Guides = exports.Guides = function Guides() {
  return null;
};
Guides.plot = _guides.guides;
var Lines = exports.Lines = function Lines() {
  return null;
};
Lines.plot = _lines.lines;
var Points = exports.Points = function Points() {
  return null;
};
Points.plot = _points.points;
var Ranges = exports.Ranges = function Ranges() {
  return null;
};
Ranges.plot = _ranges.ranges;
var Text = exports.Text = function Text() {
  return null;
};
Text.plot = _text.text;

var ScatterplotLabels = exports.ScatterplotLabels = function ScatterplotLabels() {
  return null;
};
ScatterplotLabels.plot = _scatterplotLabels.scatterplotLabels;
