// orama/lib/CanvasInput/methods/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
// Copyright 2017 Kensho Technologies, Inc.

var runHoverSolverOn = exports.runHoverSolverOn = function runHoverSolverOn(dataUnderMouse) {
  var layerProps = dataUnderMouse.layerProps,
      renderDatum = dataUnderMouse.renderDatum,
      hoverData = dataUnderMouse.hoverData,
      localMouse = dataUnderMouse.localMouse,
      mouse = dataUnderMouse.mouse;


  if (!renderDatum || !layerProps) return dataUnderMouse;
  var hoverSolver = layerProps.hoverSolver || renderDatum.hoverSolver;
  if (!hoverSolver) return dataUnderMouse;
  var hoverSolverData = hoverSolver(layerProps, hoverData, renderDatum, localMouse);
  return {
    layerProps: layerProps,
    hoverOriginalData: hoverData,
    hoverRenderData: hoverSolverData.hoverRenderData,
    hoverData: hoverSolverData.hoverData,
    renderDatum: renderDatum,
    localMouse: localMouse,
    mouse: mouse
  };
};

var getMouseFromEvt = exports.getMouseFromEvt = function getMouseFromEvt(evt) {
  if (evt.touches) {
    return {
      x: evt.touches[0].clientX,
      y: evt.touches[0].clientY
    };
  }
  return {
    x: evt.clientX,
    y: evt.clientY
  };
};
