// orama/lib/utils/path2DUtils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
// Copyright 2017 Kensho Technologies, Inc.

/**
 * Returns a Path2D object if the constructor exists, otherwise returns a mocking object.
 * This way we can keep running and testing on node code that uses Path2D
 * @namespace /utils/path
 */

/**
 * Creates a pathMock object
 * @memberOf /utils/path
 *
 * @function pathMock
 * @return {Object}
 */
var pathMock = exports.pathMock = function pathMock() {
  return {
    addPath: function addPath() {},
    closePath: function closePath() {},
    moveTo: function moveTo() {},
    lineTo: function lineTo() {},
    bezierCurveTo: function bezierCurveTo() {},
    quadraticCurveTo: function quadraticCurveTo() {},
    arc: function arc() {},
    arcTo: function arcTo() {},
    ellipse: function ellipse() {},
    rect: function rect() {}
  };
};

/**
 * According to the environments, creates a new `Path2D` object or a `pathMock`.
 * @memberOf /utils/path
 *
 * @function DEFAULT
 * @return {Path2D | pathMock}
 */
var getPath2D = exports.getPath2D = function getPath2D() {
  if (global.Path2D) return new Path2D();
  return pathMock();
};
