// orama/lib/ChartBackground/BottomLabel/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BottomLabel = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _defaultTheme = require('../../defaultTheme');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Component that position and style the bottom label of the `Chart` component
 */

var BottomLabel = exports.BottomLabel = function BottomLabel(props) {
  return _react2.default.createElement(
    'div',
    {
      style: {
        bottom: 0,
        color: props.theme.textFill,
        fontFamily: props.theme.fontFamily,
        fontSize: props.theme.axisLabelFontSize,
        fontWeight: props.theme.axisLabelFontWeight,
        left: props.plotRect.x,
        overflow: 'hidden',
        position: 'absolute',
        textAlign: 'center',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap',
        width: props.plotRect.width
      }
    },
    props.text
  );
}; // Copyright 2017 Kensho Technologies, Inc.

BottomLabel.propTypes = {
  plotRect: _propTypes2.default.object.isRequired,
  text: _propTypes2.default.string,
  theme: _propTypes2.default.object
};
BottomLabel.defaultProps = {
  plotRect: { x: 0, y: 0, width: 0, height: 0 },
  text: '',
  theme: _defaultTheme.DEFAULT_THEME
};
