// orama/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Chart = require('./Chart');

Object.defineProperty(exports, 'Chart', {
  enumerable: true,
  get: function get() {
    return _Chart.Chart;
  }
});

var _Layer = require('./Layer');

Object.keys(_Layer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _Layer[key];
    }
  });
});

var _extensions = require('./extensions');

Object.keys(_extensions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _extensions[key];
    }
  });
});

var _defaultTheme = require('./defaultTheme');

Object.defineProperty(exports, 'DEFAULT_THEME', {
  enumerable: true,
  get: function get() {
    return _defaultTheme.DEFAULT_THEME;
  }
});

var _dataGeneration = require('./utils/dataGeneration');

Object.keys(_dataGeneration).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _dataGeneration[key];
    }
  });
});

var _path2DUtils = require('./utils/path2DUtils');

Object.defineProperty(exports, 'getPath2D', {
  enumerable: true,
  get: function get() {
    return _path2DUtils.getPath2D;
  }
});

var _canvasUtils = require('./utils/canvasUtils');

Object.defineProperty(exports, 'getCachedContext', {
  enumerable: true,
  get: function get() {
    return _canvasUtils.getCachedContext;
  }
});


exports.utils = require('./utils');
exports.chartCore = require('./chartCore');
