// orama/lib/chartCore/defaults/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
// Copyright 2017 Kensho Technologies, Inc.

var ACCESSORS_NAMES = exports.ACCESSORS_NAMES = ['x', 'x0', 'x1', 'x2', 'y', 'y0', 'y1', 'y2', 'radius', 'fill', 'stroke', 'alpha', 'lineWidth', 'lineDash', 'hoverStroke', 'hoverFill'];
var ACCESSORS_NAMES_NON_SCALABLE = exports.ACCESSORS_NAMES_NON_SCALABLE = ['fillAlpha', 'strokeAlpha', 'hoverAlpha', 'hoverLineWidth', 'text', 'rotate', 'textAlign', 'textBaseline', 'textSnap', 'xOffset', 'yOffset', 'font', 'fontSize'];

var ACCESSORS_GROUPS = exports.ACCESSORS_GROUPS = {
  x: ['x', 'x0', 'x1', 'x2'],
  y: ['y', 'y0', 'y1', 'y2']
};
var ACCESSORS_TOOLTIP_ORDER = exports.ACCESSORS_TOOLTIP_ORDER = {
  y: 1,
  y0: 2,
  y1: 3,
  y2: 4,
  x: 5,
  x0: 6,
  x1: 7,
  x2: 8,
  radius: 9,
  fill: 10,
  stroke: 11,
  lineWidth: 12
};
var JS_TO_VIS_TYPE = exports.JS_TO_VIS_TYPE = {
  string: 'ordinal',
  number: 'linear',
  date: 'time'
};
var WIDTH = exports.WIDTH = 500;
var HEIGHT = exports.HEIGHT = 500;
var PROPORTION = exports.PROPORTION = 0.5;
var BACKGROUND_OFFSET = exports.BACKGROUND_OFFSET = 15;
var RANGE = exports.RANGE = [0, 1];
var DOMAIN = exports.DOMAIN = [0, 1];
var TYPE = exports.TYPE = 'linear';
var NICE = exports.NICE = false;
var TICK_COUNT = exports.TICK_COUNT = 1;
var TICK_X_SPACE = exports.TICK_X_SPACE = 150;
var TICK_Y_SPACE = exports.TICK_Y_SPACE = 75;
var SHOW_TICKS = exports.SHOW_TICKS = true;
var SHOW_LABELS = exports.SHOW_LABELS = true;
var AXIS_LABEL_OFFSET = exports.AXIS_LABEL_OFFSET = function AXIS_LABEL_OFFSET(theme) {
  return theme.axisLabelFontSize * (theme.lineHeight - 0.75);
};
var AXIS_TICK_OFFSET = exports.AXIS_TICK_OFFSET = function AXIS_TICK_OFFSET(theme) {
  return theme.axisTickFontSize * (theme.lineHeight - 0.75);
};
var PLOT_RECT = exports.PLOT_RECT = {
  x: 10, y: 10, width: 480, height: 480
};
var MARGIN_RIGHT = exports.MARGIN_RIGHT = 15;
