// orama/lib/utils/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.splitBy = exports.isDatum = exports.notDatum = exports.isPlotNumber = exports.notPlotNumber = undefined;

var _reduce3 = require('lodash/reduce');

var _reduce4 = _interopRequireDefault(_reduce3);

var _isNull2 = require('lodash/isNull');

var _isNull3 = _interopRequireDefault(_isNull2);

var _isUndefined2 = require('lodash/isUndefined');

var _isUndefined3 = _interopRequireDefault(_isUndefined2);

var _some2 = require('lodash/some');

var _some3 = _interopRequireDefault(_some2);

var _isArray2 = require('lodash/isArray');

var _isArray3 = _interopRequireDefault(_isArray2);

var _isNumber2 = require('lodash/isNumber');

var _isNumber3 = _interopRequireDefault(_isNumber2);

var _isNaN2 = require('lodash/isNaN');

var _isNaN3 = _interopRequireDefault(_isNaN2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

// Copyright 2017 Kensho Technologies, Inc.

exports.labeler = require('./labeler');
exports.canvasUtils = require('./canvasUtils');
exports.dataGeneration = require('./dataGeneration');
exports.imageRender = require('./imageRender');
exports.path2DUtils = require('./path2DUtils');
exports.rectUtils = require('./rectUtils');
exports.windowUtils = require('./windowUtils');
exports.stateHOC = require('./stateHOC');

var checkNotPlotNumber = function checkNotPlotNumber(value) {
  return (0, _isNaN3.default)(value) || !(0, _isNumber3.default)(value);
};
var notPlotNumber = exports.notPlotNumber = function notPlotNumber(value) {
  if (!(0, _isArray3.default)(value)) return checkNotPlotNumber(value);
  return (0, _some3.default)(value, checkNotPlotNumber);
};

var checkIsPlotNumber = function checkIsPlotNumber(value) {
  return !(0, _isNaN3.default)(value) && (0, _isNumber3.default)(value);
};
var isPlotNumber = exports.isPlotNumber = function isPlotNumber(value) {
  if (!(0, _isArray3.default)(value)) return checkIsPlotNumber(value);
  return (0, _some3.default)(value, checkIsPlotNumber);
};

var notDatum = exports.notDatum = function notDatum(value) {
  return (0, _isUndefined3.default)(value) || (0, _isNaN3.default)(value) || (0, _isNull3.default)(value);
};

var isDatum = exports.isDatum = function isDatum(value) {
  return !notDatum(value

  // returns (start, end] as opposed to [].slice() returning [start, end)
  );
};var slice = function slice(arr, start, end) {
  return arr.slice(start + 1, end + 1);
};

var splitBy = exports.splitBy = function splitBy(arr, iteratee) {
  var _reduce2 = (0, _reduce4.default)(arr, function (_ref, val, idx) {
    var sliceFrom = _ref.sliceFrom,
        returnArray = _ref.returnArray;

    if (iteratee(val, idx)) {
      return { sliceFrom: idx, returnArray: [].concat(_toConsumableArray(returnArray), [slice(arr, sliceFrom, idx)]) };
    }
    return { sliceFrom: sliceFrom, returnArray: returnArray };
  }, { sliceFrom: 0, returnArray: [] }),
      sliceFrom = _reduce2.sliceFrom,
      returnArray = _reduce2.returnArray;

  return [].concat(_toConsumableArray(returnArray), [slice(arr, sliceFrom, arr.length)]);
};
